core/capabilities/compute/cache.go

```diff
@@ -0,0 +1,123 @@
+package compute
+
+import (
+	"sync"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink-common/pkg/workflows/wasm/host"
+)
+
+var (
+	moduleCacheHit = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "compute_module_cache_hit",
+		Help: "hit vs non-hits of the module cache for custom compute",
+	}, []string{"hit"})
+	moduleCacheEviction = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "compute_module_cache_eviction",
+		Help: "evictions from the module cache",
+	})
+	moduleCacheAddition = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "compute_module_cache_addition",
+		Help: "additions to the module cache",
+	})
+)
+
+type moduleCache struct {
+	m  map[string]*module
+	mu sync.RWMutex
+
+	wg       sync.WaitGroup
+	stopChan services.StopChan
+
+	tickInterval time.Duration
+	timeout      time.Duration
+
+	clock    clockwork.Clock
+	onReaper chan struct{}
+}
+
+func newModuleCache(clock clockwork.Clock, tick, timeout time.Duration) *moduleCache {
+	return &moduleCache{
+		m:            map[string]*module{},
+		tickInterval: tick,
+		timeout:      timeout,
+		clock:        clock,
+		stopChan:     make(chan struct{}),
+	}
+}
+
+func (mc *moduleCache) start() {
+	mc.wg.Add(1)
+	go func() {
+		defer mc.wg.Done()
+		mc.reapLoop()
+	}()
+}
+
+func (mc *moduleCache) close() {
+	close(mc.stopChan)
+	mc.wg.Wait()
+}
+
+func (mc *moduleCache) reapLoop() {
+	ticker := mc.clock.NewTicker(mc.tickInterval)
+	for {
+		select {
+		case <-ticker.Chan():
+			mc.evictOlderThan(mc.timeout)
+			if mc.onReaper != nil {
+				mc.onReaper <- struct{}{}
+			}
+		case <-mc.stopChan:
+			return
+		}
+	}
+}
+
+func (mc *moduleCache) add(id string, mod *module) {
+	mc.mu.Lock()
+	defer mc.mu.Unlock()
+	mod.lastFetchedAt = time.Now()
+	mc.m[id] = mod
+	moduleCacheAddition.Inc()
+}
+
+func (mc *moduleCache) get(id string) (*module, bool) {
+	mc.mu.Lock()
+	defer mc.mu.Unlock()
+	gotModule, ok := mc.m[id]
+	if !ok {
+		moduleCacheHit.WithLabelValues("false").Inc()
+		return nil, false
+	}
+
+	moduleCacheHit.WithLabelValues("true").Inc()
+	gotModule.lastFetchedAt = mc.clock.Now()
+	return gotModule, true
+}
+
+func (mc *moduleCache) evictOlderThan(duration time.Duration) {
+	mc.mu.Lock()
+	defer mc.mu.Unlock()
+
```

Comment:
Not 100% required for this PR, but if it's not too much work, can you check some max size before evicting? 

If the cache isn't large, there's no point in kicking anything out.  Then we can only evict until we are also happy with the space we're consuming.

The eviction order will be random because of the nature of map, but I think that's ok for our purposes.